### PR TITLE
ServerSideRender: add new `skipBlockSupportAttributes` prop

### PR DESCRIPTION
**Pull request by t-hamano**
Related to:

- #44439
- #44438
- #43653

For more information, see the discussion that begins with [this comment](https://github.com/WordPress/gutenberg/pull/44438#pullrequestreview-1119477064).

## What?
This PR adds a new `skipBlockSupportAttributes` prop to the `ServerSideRender` component. If this prop is specified, the `ServerSideRender` component will not include attributes and styles applied by the block support in the API request data.

This will make it easier to properly add support to blocks with `ServerSideRender` components.

## Why?
When styles are applied with block support, attributes such as `attributes.textColor` and `attribites.style.border.radius` are added. If these are passed to the callback function via the `ServerSideRender` component, `get_block_wrapper_attributes` will generate the styles and CSS classes. This is the correct behavior on the front end, but on the editor, both the block wrapper and the rendered element will have duplicate styles and CSS classes.

This problem can be solved by properly excluding properties and styles added by block support from the attributes passed to the `ServerSideRender` component. However, it would make sense for components to include the logic and provide them as props.

## How?
Remove all properties and styles automatically added by the block supports from the attributes passed to the `ServerSideRender` component. This ensures that the block support styles are properly applied only to the block wrapper element on the editor.

## Testing Instructions
- Insert one of the following blocks using the server-side render component: 
    - `core/archives`
    - `core/latest-comments`
    - `core/rss`
    - `core/tag-cloud`
- In `block.json`, opt-in all block support as follows:

<details>
<summary>block.json support properties</summary>

```json
{
	"supports": {
		"align": true,
		"html": false,
		"color": {
			"text": true,
			"background": true,
			"gradients": true,
			"link": true
		},
		"typography": {
			"fontSize": true,
			"lineHeight": true,
			"__experimentalFontFamily": true,
			"__experimentalFontWeight": true,
			"__experimentalFontStyle": true,
			"__experimentalTextTransform": true,
			"__experimentalTextDecoration": true,
			"__experimentalLetterSpacing": true
		},
		"spacing": {
			"margin": true,
			"padding": true
		},
		"__experimentalBorder": {
			"radius": true,
			"color": true,
			"width": true,
			"style": true
		}
	}
}
```
</details>

- Specify the `skipBlockSupportAttributes` prop as follows:

```javascript
<ServerSideRender
	skipBlockSupportAttributes
	block="core/latest-comments"
	attributes={ attributes }
	urlQueryArgs={ { _locale: 'site' } }
/>
```
- Confirm that elements rendered by the server side are not given block support styles or classes.
    
### Before

before

### After

after

## Next step

This PR does not fix the duplicate custom CSS class that #44439, #44438, and #43653 are trying to resolve. Because I think duplicate CSS classes should be fixed regardless of this new prop. Once this PR is properly merged, I would like to submit a new PR to resolve the CSS class duplication.

Once the CSS class duplication is resolved and the skipBlockSupportAttributes are available, the following PR should be solved with simpler code:

- #44439
- #44438

However, for calendar blocks, this option would not work because color support is need to be applied to the internal elements, as was done in #42029.

- #43653

### CODE REVIEW

File: packages/server-side-render/CHANGELOG.md
Comment (t-hamano) [inline packages/server-side-render/CHANGELOG.md:7]: If the description is unnatural, please feel free to update it 🙏

**Review by aaronrobertshaw — COMMENTED** [review]
Thanks for your hard work on this @t-hamano 👍 

It is testing pretty well for me so far. I was wondering, though, if we might need to add some additional logic to `removeBlockSupportAttributes`. 

A block might wish for some block supports to be serialized on the outer wrapper while it skips other supports so they can be serialized on inner elements. In the latter case, we don't want to remove that support's attributes.

The Calendar block, as you mention, is a prime example of this. It applies typography support classes and styles to the wrapper while skipping serialization of the color support classes and styles so they can be applied internally.

Could we check each block support for `__experimentalSkipSerialization` within the block type's config? If it is set, don't remove those attributes. It is still possible for a block to skip serialization only to reapply it again on the wrapper. For such an edge case, the block itself would have to handle stripping the attributes itself. That should be pretty uncommon so I think that's fine.

<details>
<summary>Screenshot for Archives block showing no duplicate classes & styles</summary>

Screen Shot 2022-09-30 at 4 05 11 pm
</details>

**Comment by t-hamano** [comment]
Thank you for the review, @aaronrobertshaw!
I think I managed to implement the logic to check for `__experimentalSkipSerialization`.

The process of removing the block supports was very complicated. For example, in the case of text color, the following two properties need to be deleted:

- `attributes.textColor`
- `attributes.style.color.text`

For borders, it is more complicated. For example, `border-width` requires consideration of all of the following properties:

- `attributes.style.border.width`
-` attributes.style.border.top.width`
- `attributes.style.border.left.width`
- `attributes.style.border.bottom.width`
- `attributes.style.border.left.width`

At the same time, I made the following two changes:

- I separated the processing into util.js and constants.js
- Add more meaningful unit tests

I have tested in a calendar block where `__experimentalSkipSerialization` is used, and I believe that the intended rendering could be achieved by simply opting into `skipBlockSupportAttributes`.

<details>
<summary>A calendar block without skipBlockSupportAttributes</summary>

before
</details>

<details>
<summary>A calendar block with skipBlockSupportAttributes</summary>

after
</details>

I believe this new prop alone will solve the other PRs as well 🙂

- #44439
- #44438
- #43653

File: packages/server-side-render/src/constants.js
Comment (t-hamano) [inline packages/server-side-render/src/constants.js]: This definition is based on the following:
https://github.com/WordPress/gutenberg/blob/b21766438d9bb0de9c04bc2494d7093fb12b9f25/packages/blocks/src/api/constants.js#L17

File: packages/server-side-render/src/utils.js
Comment (t-hamano) [inline packages/server-side-render/src/utils.js]: This function is the same as: 
https://github.com/WordPress/gutenberg/blob/f26adefac0c5c3c52eed6689e5b8113dd11ebf51/packages/block-editor/src/hooks/utils.js#L101

**Review by aaronrobertshaw — COMMENTED** [review]
Amazing effort persevering with the complexities here @t-hamano! 🙇 

> The process of removing the block supports was very complicated.

I didn't appreciate or maybe I'd repressed memories of how tangled the block supports config and skipping serialization can get 🙈 

Knowing the effort you've already put in, I'm a bit reluctant to raise the question of whether this added complexity is worth it here. We might benefit from getting a second opinion on the direction. @andrewserong do you have any thoughts on this PR?

In case we do decide against adding the complexity to this component, you have my apologies for sending you down that path! I don't imagine it was much fun 🙂 

---

I'll outline a few of my concerns below that give me a little cause to rethink our approach.

1. There's quite a lot of code being duplicated from other packages here e.g. the constants config, `shouldSkipSerialization`, `cleanEmptyObject` etc.
2. I'm not sure people working on block supports would know of or even find the new constants etc here to update.
3. This PR introduces further usage of Lodash while there is a concerted effort in Gutenberg to remove our reliance on Lodash so that can be dropped altogether as a dependency.
4. It appears that Calendar is the only block using `ServerSideRender` that skips block support serialization. So it would be the only one benefiting from all the extra complexity added in this PR now.

Would we be better off treating Calendar as a special case and having it not leverage the new `skipBlockSupportAttributes` prop and instead stripping its own attributes as needed?

What do you think is the best path forward?

**Comment by t-hamano** [comment]
I don't care about the additional implementation regarding serialization, as I have no problem if this is not incorporated 👍

I'm rethinking my thinking, it might be a rare case for blocks with the `ServerSideRender` component to skip serialization, like a calendar block? 🤔 If so, this additional implementation might not be beneficial enough to cover the concerns you listed (Use of lodash, duplication of functions and config, maintenance of config, etc).

**Review by andrewserong — COMMENTED** [review]
Just echoing some of @aaronrobertshaw's thoughts here, but thank you so much @t-hamano for going down this particular rabbithole. I think it's very helpful for revealing some of the trade-offs of the decision making here surrounding both how we use the `ServerSideRender` component, and where should we put the complexity.

From my perspective, I think the scope of this PR now reveals that we might be better served to deal with the additional duplication in the original PRs that dealt with the block supports problem on a case-by-case basis for each block, in the block's edit components. The reason being that the complexity added here also winds up introducing duplication, that would then need to be maintained as the block supports evolve. Plus, there is also the concern that ideally in the long-term, blocks will reduce their dependency on using the `ServerSideRender` component altogether.

So, it might wind up being a bit easier for us if we deal with this specifically in the blocks' Edit components as in the original PRs (e.g https://github.com/WordPress/gutenberg/pull/44438)?

I very much share @aaronrobertshaw's sentiment that this is great work here, I think it's hugely valuable to explore all our options in depth, and it's very much appreciated! It's always challenging with these sorts of problems where there isn't a single obvious, clean solution, but rather a set of trade-offs to manage complexity and maintainability.

What do you both think?

**Comment by aaronrobertshaw** [comment]
Thanks for the sanity check @andrewserong, appreciate it.

> I think the scope of this PR now reveals that we might be better served to deal with the additional duplication in the original PRs that dealt with the block supports problem on a case-by-case basis for each block, in the block's edit components

My thinking was that we could revert this PR to the [earlier state](https://github.com/WordPress/gutenberg/blob/b70814f34696dc94cf8572b688d7cfdc7eedeb5a/packages/server-side-render/src/server-side-render.js), where it just performs a blanket removal of all block support attributes if the new prop is set. This would work for the majority of cases leaving only the Calendar block needing a bespoke solution in its edit component yet still avoiding most of the code duplication.

I'm happy for @t-hamano to decide what's best and requires the least re-work for him.

**Comment by andrewserong** [comment]
> My thinking was that we could revert this PR to the [earlier state](https://github.com/WordPress/gutenberg/blob/b70814f34696dc94cf8572b688d7cfdc7eedeb5a/packages/server-side-render/src/server-side-render.js), where it just performs a blanket removal of all block support attributes if the new prop is set.

Good point, I think that would be very reasonable, too, as it's only a minimal addition to the API for the `ServerSideRender` component, and the logic is pretty contained.

> I'm happy for @t-hamano to decide what's best and requires the least re-work for him.

Sounds good 👍

**Comment by t-hamano** [comment]
Thanks for the input, @aaronrobertshaw, @andrewserong !

> My thinking was that we could revert this PR to the [earlier state](https://github.com/WordPress/gutenberg/blob/b70814f34696dc94cf8572b688d7cfdc7eedeb5a/packages/server-side-render/src/server-side-render.js), where it just performs a blanket removal of all block support attributes if the new prop is set.

I, too, think it is appropriate to land at this state, but it is a difficult question to answer when considering the future.
Certainly, it is useful to be able to control attributes about block support with this prop alone, but if block support increases in the future, this [hard-coded section](https://github.com/WordPress/gutenberg/blob/b70814f34696dc94cf8572b688d7cfdc7eedeb5a/packages/server-side-render/src/server-side-render.js#L28-L39) will need to be maintained. This is my only concern.

As a final decision for me, I think this prop isn't necessary now, and how about control attributes in each block? In other words, I think it's a good idea to close this PR and proceed with the following PRs with individual fixes in each block.

- #44439
- #44438
- #43653
- 
There are only five core blocks where the `ServerSideRender` component was used, and there will not be more in the future.

This decision could not have been reached had we not actually worked on the development, and I believe that this PR is not in vain 🙂

What do you think about this opinion?

**Comment by aaronrobertshaw** [comment]
> if block support increases in the future, this [hard-coded section](https://github.com/WordPress/gutenberg/blob/b70814f34696dc94cf8572b688d7cfdc7eedeb5a/packages/server-side-render/src/server-side-render.js#L28-L39) will need to be maintained. This is my only concern.

Another way to look at this is that if block supports change in future you'd still likely have to update each of the individual blocks which is still more maintenance than updating this component.

New supports are also still being added to these blocks as well e.g. [typography for Tag Cloud](https://github.com/WordPress/gutenberg/pull/43452). 

Given that I'm starting to lean more to the middle ground option I [proposed earlier](https://github.com/WordPress/gutenberg/pull/44491#issuecomment-1267799269).

**Comment by andrewserong** [comment]
> There are only five core blocks where the ServerSideRender component was used, and there will not be more in the future.

> Given that I'm starting to lean more to the middle ground option I https://github.com/WordPress/gutenberg/pull/44491#issuecomment-1267799269.

You both make compelling arguments! I'd be pretty happy with either outcome, I think.

**Comment by t-hamano** [comment]
> Another way to look at this is that if block supports change in future you'd still likely have to update each of the individual blocks which is still more maintenance than updating this component.

I see. no matter which approach we choose, maintenance will be always needed...😥
Then, after all, shall we let it land in [this earlier state](https://github.com/WordPress/gutenberg/blob/b70814f34696dc94cf8572b688d7cfdc7eedeb5a/packages/server-side-render/src/server-side-render.js) (Just remove all block support attributes if `skipBlockSupportAttributes` is set)?

**Comment by aaronrobertshaw** [comment]
Keeping the maintenance burden to a minimum sounds like a solid plan. 👍 

An added bonus might be keeping those other blocks simpler so there's one thing less to grok when it comes to rewriting them to no longer rely on `ServerSideRender`.

**Comment by t-hamano** [comment]
✅ I have reverted some commits and reverted back to [this state](https://github.com/WordPress/gutenberg/blob/b70814f34696dc94cf8572b688d7cfdc7eedeb5a/packages/server-side-render/src/server-side-render.js) that doesn't check serialization status.
✅ The latest trunk was taken in again.
✅ I forgot to remove duplicate `className`, so I added that process on 7b02d8d.
✅ Updated unit test on 4be4688 to confirm that `className` is correctly removed.

The Testing Instructions described in the first comment of this PR should produce the expected results as before 👍

File: packages/server-side-render/README.md
Comment (aaronrobertshaw) [inline packages/server-side-render/README.md]: Could it be worth adding a quick note here that this will remove all attributes and style properties regardless of whether certain features skip serialization of block support styles?

File: packages/server-side-render/README.md
Comment (aaronrobertshaw) [inline packages/server-side-render/README.md:85]:
```suggestion
-   Required: No
-   Default: false
```

I believe we usually want to note a prop's default value when it has one in these READMEs. If we add this we might as well tweak the `httpMethod` entry as well.

**Review by aaronrobertshaw — APPROVED** [review]
Appreciate your persistence in iterating on this one @t-hamano 🙇 

It is testing well for me. The unit tests pass and all duplicate classes and styles were removed for the following blocks using `ServerSideRender`.

✅ Archives
✅ Latest Comments
✅ RSS
✅ Tag Cloud

I left a couple of inline comments to tweak the readme. Other than those suggestions, I'd say this is ready to go. Nice work!

File: packages/server-side-render/README.md
Comment (t-hamano) [reply packages/server-side-render/README.md]: Updated by d58e60c
If this description is sufficient I would like to merge 👍

Comment (aaronrobertshaw) [reply packages/server-side-render/README.md]: Thanks, that works for me. 🚢

**Comment by bph** [comment]
Although only four core/blocks benefit from the change in the ServerSideRender package, [there are 620 plugins in the repo](https://wpdirectory.net/search/01GGCN803Q5WMG2RFBY2CNZMPA) who also use it one way or other.  

Would I be correct, when I write: 

Before this is merged into Core with 6.2 
- Are there any backwards-compatibility concerns, that need to be considered? 
- stabilize of `__experimentalSkipSerialization`
- write a devnote elaborating on this description: 
_"Remove attributes and style properties applied by the block supports. This prevents duplication of styles in the block wrapper and the `ServerSideRender` components. Even if certain features skip serialization to HTML markup by `__experimentalSkipSerialization`, all attributes and style properties are removed."_

**Comment by t-hamano** [comment]
@bph 

> Are there any backwards-compatibility concerns, that need to be considered?

This new property will keep the existing behavior unless explicitly opted in by the developer. Therefore, I think that the backward compatibility doesn't need to be considered.

> stabilize of __experimentalSkipSerialization

Perhaps this might need to be discussed in #45194; whether to stabilize `__experimentalSkipSerialization` is also discussed.

> write a devnote elaborating on this description:

I am wondering if a dev note should be written on this.

This new prop is implemented to solve the problem of double application of block support styles in core blocks using `ServerSideRender`. Of course, this new prop can also be used by plugin developers, but the use of this component itself is currently discouraged.

Therefore, it might not be necessary to actively announce it.

@aaronrobertshaw 
Please let us know if you have any feedback on this matter.

**Comment by bph** [comment]
Thanks @t-hamano for entertaining my questions.
> the use of this component itself is currently discouraged.

Ah interesting, was this formulized in any documentation? 620 plugins can't be that wrong :-)

**Comment by t-hamano** [comment]
Please see: https://github.com/WordPress/gutenberg/tree/trunk/packages/server-side-render#serversiderender

My understanding is that this package used to be used to render dynamic blocks on the editor. And it had the advantage of being able to use the PHP rendering logic as it is done on the front end. However, nowadays, with the data API, it is possible to implement rendering on the editor side in JavaScript without using this Package.

Many developers may continue to use this package because it eliminates the need to write a lot of editor code 🤔

**Comment by aaronrobertshaw** [comment]
> Please let us know if you have any feedback on this matter.

Thanks for the ping 👍 

I think @t-hamano sums everything up nicely.

The only point I would add relates to the stabilization of the `__experimentalSkipSerialization` API. It has actually come up on https://github.com/WordPress/gutenberg/issues/45194. In addition to that, there are plans to stabilize all the block support-related APIs, which would include this one for skipping their serialization. You can see this as Phase 3 in the issue [tracking design tool consistency](https://github.com/WordPress/gutenberg/issues/43241).

For the moment, some other Gutenberg Phase 2 wrap-up tasks are taking priority, but the hope is to address those stabilizations soon.

**Comment by bph** [comment]
>My understanding is that this package used to be used to render dynamic blocks on the editor. And it had the advantage of being able to use the PHP rendering logic as it is done on the front end. However, nowadays, with the data API, it is possible to implement rendering on the editor side in JavaScript without using this Package.

Given that Gutenberg is six years old, the group of PHP developer using `ServerSideRender`, is not surprising. What would you relay to the extenders who are using the `ServerSideRender` as to what changed? They would really appreciate a heads-up. 

>This new property will keep the existing behavior unless explicitly opted in by the developer. Therefore, I think that the backward compatibility doesn't need to be considered.

That's good to know. So this shouldn't be more than a two paragraph information, as to the why, what, how?

**Comment by t-hamano** [comment]
I am wondering if we should put this PR in the dev note, I have two ideas:

- Include in the dev note. However, as noted in the component documentation, I think it would be better to include an additional description that new blocks should be written in JavaScript instead. Of course, I would be happy to write the dev note 👍
- Just add a code example about [this attribute](https://github.com/WordPress/gutenberg/blob/trunk/packages/server-side-render/README.md#skipblocksupportattributes) in the component's README.

@aaronrobertshaw If you don't mind, I'd like to know your opinion 🙏

**Comment by aaronrobertshaw** [comment]
> If you don't mind, I'd like to know your opinion 🙏

I don't think there is much harm in communicating the change. 

For those already using `ServerSideRender` and wishing to adopt block supports without re-writing their blocks, knowing about the new prop would benefit them. We can still stress all new blocks should leverage the data API instead of `ServerSideRender`.

**Comment by t-hamano** [comment]
Thank you @aaronrobertshaw!

@bph

I have written a dev note here. Feel free to edit it.

---

# Add new prop to ServerSideRender component

WordPress 6.2 introduces a new `skipBlockSupportAttributes` prop to exclude attributes and styles related to block supports in the `ServerSideRender` component. This will make it easier to properly add support to blocks with the `ServerSideRender` components.

Note: `ServerSideRender` component should be regarded as a fallback or legacy mechanism, it is not appropriate for developing new features against. New blocks should be built in conjunction with any necessary REST API endpoints, so that JavaScript can be used for rendering client-side in the edit function. This gives the best user experience, instead of relying on using the PHP render_callback. The logic necessary for rendering should be included in the endpoint, so that both the client-side JavaScript and server-side PHP logic should require a minimal amount of differences.

`ServerSideRender` is a component used for server-side rendering a preview of dynamic blocks to display in the editor. By passing the `attributes` prop to this component, it is processed on the server side and the block receives the rendered result.

```javascript
import ServerSideRender from '@wordpress/server-side-render';
const MyServerSideRender = () => (
	<ServerSideRender
		block="core/archives"
		attributes={ attributes }
	/>
);
```

If you add block supports to the block that contains this component,  the problem is that the styles from the block support will be applied to both the block wrapper element and the rendered HTML by `ServerSiderender` component. For example, if you opt for `padding` and `margin` as block support, and those styles are applied to the block, you will have double spaces.

```html
<div
	...
	class="block-editor-block-list__block wp-block wp-block-archives"
	style="padding: 10px; margin: 10px;"
>
	<div inert="true" class="components-disabled">
  		<div>
  			<ul
				style="padding-top:10px;padding-right:10px;padding-bottom:10px;padding-left:10px;margin-top:10px;margin-right:10px;margin-bottom:10px;margin-left:10px;"
				class="wp-block-archives-list wp-block-archives"
			>
				<li><a href="http://example.com">Hello World</a></li>
			</ul>
		</div>
	</div>
</div>
```

By opting in to [the new `skipBlockSupportAttributes` prop](https://github.com/WordPress/gutenberg/pull/44491) introduced in WordPress 6.2, all attributes related to block support will be removed before rendering on the server side, and you will be able to get proper rendering results.

```javascript
import ServerSideRender from '@wordpress/server-side-render';
const MyServerSideRender = () => (
	<ServerSideRender
 		skipBlockSupportAttributes
		block="core/archives"
		attributes={ attributes }
	/>
);
```

However, if block styles are already defined in the global style or in `theme.json`, those styles will take precedence and individual block's overrides may not be applied. In such cases, explicitly handling the block support to be passed to the `ServerSideRender` component will produce the intended result.

```javascript
import ServerSideRender from '@wordpress/server-side-render';

const serverSideAttributes = {
	...attributes,
	style: {
		...attributes?.style,
		// Ignore styles related to margin and padding.
		spacing: undefined,
	},
};

const MyServerSideRender = () => (
	<ServerSideRender
		// All block support attributes except margins and padding are passed.
 		attributes={ serverSideAttributes }
		block="core/archives"
		attributes={ attributes }
	/>
);
```

**Comment by aaronrobertshaw** [comment]
@t-hamano One thing that might be worth highlighting for the dev note, is that the primary block supports that are negatively impacted by the duplication of classes and styles are the spacing supports (padding, margin). 

Some blocks have Global Styles generated under their default class e.g. `.wp-block-<name>`. If the block support attributes corresponding to those global styles aren't passed through to the `ServerSideRender` component you may end up with a situation where the global styles or theme.json styles are shown in the editor not the individual block's overrides.

I encountered the above when rebasing the PR [adopting typography supports for the Latest Comments](https://github.com/WordPress/gutenberg/pull/43310) block.

**Comment by t-hamano** [comment]
Thank you for the advice, @aaronrobertshaw!

> For example, if you opt for padding and margin as block support, and those styles are applied to the block, you will have double spaces.

I have added the above text and updated the code examples slightly.

> Some blocks have Global Styles generated under their default class e.g. .wp-block-<name>. If the block support attributes corresponding to those global styles aren't passed through to the ServerSideRender component you may end up with a situation where the global styles or theme.json styles are shown in the editor not the individual block's overrides.

My understanding is that such a case occurs when "__experimentalSelector" or `__experimentalSkipSerialization` is defined. I think it isn't necessary to communicate concerns about experimental APIs to third party developers, how do you think?

**Comment by aaronrobertshaw** [comment]
> My understanding is that such a case occurs when "__experimentalSelector" or __experimentalSkipSerialization is defined.

This isn't the case, unfortunately. 

Using https://github.com/WordPress/gutenberg/pull/43310 as an example. It doesn't define any custom selector and doesn't skip serialization.

If you set global styles for the Latest Comments block's typography, then switch to the editor and try to change these typography styles for a single Latest Comments block, they wouldn't be applied in the editor unless we still allow those individual block typography styles to be applied to the ServerSideRendered markup. 

It's this fact that led me to need to remove the `skipBlockSupportAttributes` from the LatestComments block and instead filter the attributes passed to the ServerSideRender component.

Given that example doesn't relate to experimental APIs such as `__experimentalSelector` or `__experimentalSkipSerialization`, I think it is worth noting.

**Comment by t-hamano** [comment]
Okay, I understand.
So you are saying that such a problem would occur if the styles related to the block support are already applied to elements within the block?

> However, if block styles are already defined in the global style or in theme.json, those styles will take precedence and individual block's overrides may not be applied. In such cases, explicitly handling the block support to be passed to the ServerSideRender component will produce the intended result.

I have added the above instructions and a new code example. Feel free to update if there is anything that is not clear 🙏

**Comment by aaronrobertshaw** [comment]
> So you are saying that such a problem would occur if the styles related to the block support are already applied to elements within the block?

I'll try and clarify this further, still using the latest comments block example;

In the editor, the block's wrapper gets the `wp-block-latest-comments` class, but so does the markup returned by the `ServerSideRender` component. This means we end up with something like the following:

Screenshot 2023-02-10 at 2 18 40 pm

If you set, say font family, on the block in Global Styles, the global styles stylesheet gets the following:
`.wp-block-latest-comments { font-family: 'my-custom-font'; }`. That would be applied to both elements with that class in the above screenshot.

If you then apply a new font family on an individual block in the post editor, that gets applied to the block's outer wrapper. If you also pass `skipBlockSupportAttributes` to the `ServerSideRender` component, that individual block's style isn't applied on the inner `wp-block-latest-comments` element and the CSS from the global styles stylesheet would override the inheritance from the outer wrapper.

I hope that makes things a little clearer 🤞

**Comment by t-hamano** [comment]
Thanks for the detailed explanation, @aaronrobertshaw! I think I have understood it well enough 🙏

**Comment by bph** [comment]
@t-hamano Thanks for going through the process and finalizing the Dev Note. Would you please add this to the Make blog as draft so we can have it reviewed for publishing?  Please let me know if you need author access to the site.

**Comment by t-hamano** [comment]
@bph 

I have written a dev note draft: https://make.wordpress.org/core/?p=102933

**Comment by mcsf** [comment]
Hi! 👋 I'm catching up on this PR after reading the [dev note](https://make.wordpress.org/core/2023/03/06/add-new-prop-to-serversiderender-component/).

One thing that struck me is that this setup with `removeBlockSupportAttributes` may easily get out of sync: if a new feature is added to `lib/block-supports`, there is nothing to remind the developers to update `removeBlockSupportAttributes`.

Even as we speak, there are discrepancies between that function and the actual map of support attributes:

```sh
# List all attributes names corresponding to block-supports features:
$ grep -Eho "attributes\['\w+'\]" lib/block-supports/*.php | awk -F\' '{print $2}' | sort | uniq

anchor
backgroundColor
borderColor
class
fontFamily
fontSize
gradient
layout
shadow
style
textColor
```

If we treat `className` and `style` as special attributes, this still leaves two keys out of `removeBlockSupportAttributes`: `anchor` and `layout`. How can we properly keep track of what attributes should be removed? How do we prevent future problems?

**Comment by t-hamano** [comment]
@mcsf
Thanks for letting me know.

We discussed in this PR whether we should finally introduce the `skipBlockSupportAttributes` prop.

As you say, when new block support is added, its attributes and styles should be added in the `ServerSideRender` component as well. Conversely, if this prop were not in place, then each time we add new support to a block rendered by the `SeverSideRender` component, we would need to properly exclude attributes and styles in each block.

Maintenance would occur either way, but after consideration, I made a decision that introducing this prop would be less costly to maintain.

As far as I can see, the new properties that should be added as exclusions are the `anchor` and `shadow` attributes. If these properties need to be excluded (and if it is possible to backport to WordPress 6.2 now), I would be happy to accommodate.

In any case, we may want to explore whether there is a way to detect when new block support is added in the future and update this package automatically or manually.

cc: @aaronrobertshaw, @andrewserong

**Comment by t-hamano** [comment]
As far as I know, the `shadow` property is not supported at the block level and can only be controlled by the global style. Therefore, no block-specific attributes are generated from this property.

I added the `shadow` property to an archive block with a `ServerSideRender` component as a test. Yes, styles are applied in duplicate, but this is due to inline styles generated from global styles, which are not controllable by the `ServerSideRender` component.

To prevent this, it is generally necessary to control this with the `__experimentalSkipSerialization` or `__experimentalSelector `property. However, since the shadow property only allows boolean types, it appears that these cannot be addressed. Therefore, as far as the shadow property is concerned, this is not a problem that can be handled by the `ServerSideRender` component, and I believe that it needs to be addressed in some way by the block support itself.

shadow

**Pull request merged**
